File: engines/python/inserter_pgsql.py
import csv
import psycopg2
from psycopg2.extras import execute_values


def _decode_libpq_bytes(raw: bytes) -> str:
    """Перебор вероятных кодировок ошибок libpq (см. src/db/pgsql.py)."""
    for enc in ("utf-8", "cp1251", "cp1252", "latin-1"):
        try:
            return raw.decode(enc)
        except UnicodeDecodeError:
            continue
    return raw.decode("latin-1", errors="replace")


def _connect(conn_params: dict):
    try:
        return psycopg2.connect(
            host=conn_params["host"],
            port=conn_params["port"],
            user=conn_params["user"],
            password=conn_params["password"],
            dbname=conn_params["database"],
            # ASCII-сообщения от сервера (применяется ПОСЛЕ авторизации)
            # и UTF-8 для данных. Для ошибок этапа подключения дополнительно
            # ловим UnicodeDecodeError ниже.
            options="-c lc_messages=C",
            client_encoding="UTF8",
        )
    except UnicodeDecodeError as e:
        raw = e.object if isinstance(e.object, (bytes, bytearray)) else b""
        msg = (
            _decode_libpq_bytes(bytes(raw)).strip()
            if raw
            else "(не удалось извлечь сообщение)"
        )
        raise RuntimeError(f"PostgreSQL connection failed: {msg}") from e


def _quote(name: str) -> str:
    clean = name.strip().strip('"').replace('"', '""')
    return f'"{clean}"'


def count_rows(conn_params: dict, table_name: str) -> int:
    """Возвращает фактическое число строк в таблице из самой БД.
    Используется в main.py для независимой проверки результата вставки."""
    conn = _connect(conn_params)
    cursor = conn.cursor()
    try:
        cursor.execute(f"SELECT COUNT(*) FROM {_quote(table_name)}")
        row = cursor.fetchone()
        return int(row[0]) if row else 0
    finally:
        cursor.close()
        conn.close()


def default_insert(conn_params: dict, csv_file: str, table_name: str) -> int:
    conn = _connect(conn_params)
    cursor = conn.cursor()
    count = 0
    try:
        with open(csv_file, "r", newline="", encoding="utf-8") as f:
            for row in csv.DictReader(f):
                col_names = ", ".join(_quote(c) for c in row.keys())
                placeholders = ", ".join(["%s"] * len(row))
                cursor.execute(
                    f"INSERT INTO {_quote(table_name)} ({col_names}) VALUES ({placeholders})",
                    list(row.values()),
                )
                count += 1
        conn.commit()
        return count
    except Exception:
        conn.rollback()
        raise
    finally:
        cursor.close()
        conn.close()


def bulk_insert(
    conn_params: dict, csv_file: str, table_name: str, batch_size: int = 1000
) -> int:
    conn = _connect(conn_params)
    cursor = conn.cursor()
    count = 0
    try:
        with open(csv_file, "r", newline="", encoding="utf-8") as f:
            reader = csv.DictReader(f)
            col_names = ", ".join(_quote(c) for c in reader.fieldnames)
            sql = f"INSERT INTO {_quote(table_name)} ({col_names}) VALUES %s"

            batch = []
            for row in reader:
                batch.append(list(row.values()))
                if len(batch) >= batch_size:
                    execute_values(cursor, sql, batch, page_size=batch_size)
                    count += len(batch)
                    batch.clear()
            if batch:
                execute_values(cursor, sql, batch, page_size=batch_size)
                count += len(batch)

        conn.commit()
        return count
    except Exception:
        conn.rollback()
        raise
    finally:
        cursor.close()
        conn.close()


def file_insert(conn_params: dict, csv_file: str, table_name: str) -> int:
    conn = _connect(conn_params)
    cursor = conn.cursor()
    try:
        with open(csv_file, "r", newline="", encoding="utf-8") as f:
            row_count = sum(1 for _ in csv.DictReader(f))

        with open(csv_file, "r", newline="", encoding="utf-8") as f:
            cursor.copy_expert(
                f"COPY {_quote(table_name)} FROM STDIN WITH (FORMAT csv, HEADER true)",
                f,
            )
        conn.commit()
        return row_count
    except Exception:
        conn.rollback()
        raise
    finally:
        cursor.close()
        conn.close()

File: engines/java/src/main/java/com/dbloadbench/InserterPgSQL.java
import org.postgresql.copy.CopyManager;
import org.postgresql.core.BaseConnection;

import java.io.*;
import java.sql.*;
import java.util.*;

public class InserterPgSQL implements Inserter {

    private final Connection conn;
    private final ConnParams params;

    public InserterPgSQL(ConnParams p) throws SQLException {
        try {
            Class.forName("org.postgresql.Driver");
        } catch (ClassNotFoundException e) {
            throw new SQLException("PostgreSQL driver not found", e);
        }
        
        String url = String.format(
            "jdbc:postgresql://%s:%d/%s?sslmode=disable",
            p.host, p.port, p.database
        );
        Properties props = new Properties();
        props.setProperty("user",     p.user);
        props.setProperty("password", p.password);

        conn         = DriverManager.getConnection(url, props);
        this.params  = p;
    }

    @Override
    public void close() {
        try { if (conn != null) conn.close(); }
        catch (SQLException ignored) {}
    }

    private String quote(String name) {
        String clean = CSVReader.cleanIdentifier(name)
                                .replace("\"", "\"\"");
        return "\"" + clean + "\"";
    }

    private String placeholder(int i) {
        return "?";
    }

    @Override
    public int defaultInsert(String csvFile, String tableName) throws Exception {
        int total = 0;
        conn.setAutoCommit(false);
        try (CSVStream csv = new CSVStream(csvFile)) {
            String cols = buildCols(csv.headers);
            String phs  = buildPlaceholders(csv.headers.size());
            String sql  = String.format("INSERT INTO %s (%s) VALUES (%s)",
                                        quote(tableName), cols, phs);

            try (PreparedStatement ps = conn.prepareStatement(sql)) {
                while (csv.hasNext()) {
                    String[] row = csv.next();
                    for (int i = 0; i < row.length; i++) {
                        ps.setString(i + 1, row[i]);
                    }
                    ps.executeUpdate();
                    total++;
                }
                conn.commit();
            }
        } catch (Exception e) {
            conn.rollback();
            throw e;
        } finally {
            conn.setAutoCommit(true);
        }
        return total;
    }

    @Override
    public int bulkInsert(String csvFile, String tableName, int batchSize) throws Exception {
        int total = 0;
        conn.setAutoCommit(false);
        try (CSVStream csv = new CSVStream(csvFile)) {
            String cols = buildCols(csv.headers);
            String phs  = buildPlaceholders(csv.headers.size());
            String sql  = String.format("INSERT INTO %s (%s) VALUES (%s)",
                                        quote(tableName), cols, phs);

            try (PreparedStatement ps = conn.prepareStatement(sql)) {
                int count = 0;
                while (csv.hasNext()) {
                    String[] row = csv.next();
                    for (int i = 0; i < row.length; i++) {
                        ps.setString(i + 1, row[i]);
                    }
                    ps.addBatch();
                    count++;

                    if (count >= batchSize) {
                        ps.executeBatch();
                        total += count;
                        count = 0;
                    }
                }
                if (count > 0) {
                    ps.executeBatch();
                    total += count;
                }
                conn.commit();
            }
        } catch (Exception e) {
            conn.rollback();
            throw e;
        } finally {
            conn.setAutoCommit(true);
        }
        return total;
    }

    @Override
    public int fileInsert(String csvFile, String tableName) throws Exception {
        // Стримим файл напрямую в PostgreSQL COPY, без парсинга в Java.
        // Парсинг и StringBuilder-копия съедали ~2× размер файла в куче и
        // приводили к OutOfMemoryError на больших объёмах (10⁷ строк).
        // PostgreSQL COPY понимает RFC 4180 CSV сам.
        String copySQL = String.format(
            "COPY %s FROM STDIN WITH (FORMAT csv, HEADER true)",
            quote(tableName)
        );

        long rowsCopied;
        conn.setAutoCommit(false);
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(new FileInputStream(csvFile), "UTF-8"),
                1 << 20  // 1 МБ буфер чтения
        )) {
            CopyManager copyManager = new CopyManager((BaseConnection) conn);
            rowsCopied = copyManager.copyIn(copySQL, reader);
            conn.commit();
        } catch (Exception e) {
            conn.rollback();
            throw e;
        } finally {
            conn.setAutoCommit(true);
        }

        return (int) rowsCopied;
    }

    private String buildCols(List<String> headers) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < headers.size(); i++) {
            if (i > 0) sb.append(", ");
            sb.append(quote(headers.get(i)));
        }
        return sb.toString();
    }

    private String buildPlaceholders(int count) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < count; i++) {
            if (i > 0) sb.append(", ");
            sb.append(placeholder(i));
        }
        return sb.toString();
    }

    private String escapeCsvField(String value) {
        if (value.contains(",") || value.contains("\"") || value.contains("\n")) {
            return "\"" + value.replace("\"", "\"\"") + "\"";
        }
        return value;
    }
}